cms/lib/xblock/upstream_sync.py

```diff
@@ -0,0 +1,451 @@
+                version_declined=None,
+                error_message=str(exc),
+            )
+
+    @classmethod
+    def get_for_block(cls, downstream: XBlock) -> t.Self:
+        """
+        Get info on a block's relationship with its linked upstream content (without actually loading the content).
+
+        Currently, the only supported upstreams are LC-backed Library Components. This may change in the future (see
+        module docstring).
+
+        If link exists, is supported, and is followable, returns UpstreamLink.
+        Otherwise, raises an UpstreamLinkException.
+        """
+        if not downstream.upstream:
+            raise NoUpstream()
+        if not isinstance(downstream.usage_key.context_key, CourseKey):
+            raise BadDownstream(_("Cannot update content because it does not belong to a course."))
+        if downstream.has_children:
+            raise BadDownstream(_("Updating content with children is not yet supported."))
+        try:
+            upstream_key = LibraryUsageLocatorV2.from_string(downstream.upstream)
+        except InvalidKeyError as exc:
+            raise BadUpstream(_("Reference to linked library item is malformed")) from exc
+        downstream_type = downstream.usage_key.block_type
+        if upstream_key.block_type != downstream_type:
+            # Note: Currently, we strictly enforce that the downstream and upstream block_types must exactly match.
+            #       It could be reasonable to relax this requirement in the future if there's product need for it.
+            #       For example, there's no reason that a StaticTabBlock couldn't take updates from an HtmlBlock.
+            raise BadUpstream(
+                _("Content type mismatch: {downstream_type} cannot be linked to {upstream_type}.").format(
+                    downstream_type=downstream_type, upstream_type=upstream_key.block_type
+                )
+            ) from TypeError(
+                f"downstream block '{downstream.usage_key}' is linked to "
+                f"upstream block of different type '{upstream_key}'"
+            )
+        # We import this here b/c UpstreamSyncMixin is used by cms/envs, which loads before the djangoapps are ready.
+        from openedx.core.djangoapps.content_libraries.api import (
+            get_library_block  # pylint: disable=wrong-import-order
+        )
+        try:
+            lib_meta = get_library_block(upstream_key)
+        except XBlockNotFoundError as exc:
+            raise BadUpstream(_("Linked library item was not found in the system")) from exc
+        return cls(
+            upstream_ref=downstream.upstream,
+            version_synced=downstream.upstream_version,
+            version_available=(lib_meta.draft_version_num if lib_meta else None),
+            # TODO: Previous line is wrong. It should use the published version instead, but the
+            # LearningCoreXBlockRuntime APIs do not yet support published content yet.
```

Comment:
They do now :) (mostly)

Reply:
Nice! One of us can do a follow-up right after this merges, then.